<div style="text-align: center;">
  <img src="Images/Placement.png" alt="Placement Illustration" width="600"/>
</div>

## Hello!
This is a project utilizing the **Placement Prediction Dataset** to analyze and predict student placement outcomes. By exploring this dataset, we aim to identify key factors that influence placement success and develop machine learning models for accurate classification of placement status.  

The dataset, available on [Kaggle](https://www.kaggle.com/datasets/ruchikakumbhar/placement-prediction-dataset), contains **10,000 rows** and **12 columns**, providing detailed academic and extracurricular records of students, along with their placement status. It is well-structured with no missing values, making it suitable for analysis and classification tasks. 

#### Dataset Overview
The dataset includes the following features:  

1. **CGPA**: Cumulative Grade Point Average, representing a student’s overall academic performance.  
2. **Internships**: The number of internships completed by a student.  
3. **Projects**: Number of projects undertaken by the student.  
4. **Workshops/Certifications**: Participation in workshops and online courses for skill enhancement.  
5. **Aptitude Test Score**: Measures a student’s quantitative and logical reasoning skills, often used in recruitment.  
6. **Soft Skills Rating**: Assesses communication and interpersonal skills.  
7. **Extracurricular Activities**: Indicates student involvement in non-academic activities.  
8. **Placement Training**: Training programs provided by colleges to help students prepare for placement.  
9. **SSC & HSC Marks**: Academic performance in Senior Secondary (SSC) and Higher Secondary (HSC) levels.  
10. **Placement Status**: The target variable, indicating whether a student was **Placed** or **Not Placed**.  

#### Project Inspiration  
This project is inspired by the need to analyze key factors affecting student placements and build predictive models to assist career guidance efforts. Understanding the most influential attributes can help students and institutions improve placement strategies.  

#### Goals of the Project  
1. **Data Analysis**: Explore the dataset to identify trends and correlations among academic, training, and extracurricular features.  
2. **Classification Models**: Develop and evaluate machine learning models to predict placement status based on the provided features.  

#### Dataset Overview
The dataset includes the following features:  

1. **CGPA**: Cumulative Grade Point Average, representing a student’s overall academic performance.  
2. **Internships**: The number of internships completed by a student.  
3. **Projects**: Number of projects undertaken by the student.  
4. **Workshops/Certifications**: Participation in workshops and online courses for skill enhancement.  
5. **Aptitude Test Score**: Measures a student’s quantitative and logical reasoning skills, often used in recruitment.  
6. **Soft Skills Rating**: Assesses communication and interpersonal skills.  
7. **Extracurricular Activities**: Indicates student involvement in non-academic activities.  
8. **Placement Training**: Training programs provided by colleges to help students prepare for placement.  
9. **SSC & HSC Marks**: Academic performance in Senior Secondary (SSC) and Higher Secondary (HSC) levels.  
10. **Placement Status**: The target variable, indicating whether a student was **Placed** or **Not Placed**.  

#### Project Inspiration  
This project is inspired by the need to analyze key factors affecting student placements and build predictive models to assist career guidance efforts. Understanding the most influential attributes can help students and institutions improve placement strategies.  

#### Goals of the Project  
1. **Data Analysis**: Explore the dataset to identify trends and correlations among academic, training, and extracurricular features.  
2. **Classification Models**: Develop and evaluate machine learning models to predict placement status based on the provided features.  

> ⚠️ *Disclaimer*: The goal is to build predictive tools and derive insights—not to stereotype or oversimplify student capabilities. Ethical analysis and transparency are essential.

# **Step 3: Exploratory Data Analysis (EDA) – SQL Queries**

This notebook focuses on performing **exploratory data analysis using SQL queries** within a Jupyter environment. By using SQL-style queries with the help of `sqlite3`, we can analyze and extract insights from the Placement dataset in a more intuitive and readable way for those familiar with SQL.

---

### Objectives of This Notebook

1. [Import Libraries and Load the Dataset](#import)  
2. [Preview the Dataset](#preview)  
3. [Run SQL Queries to Answer Key Questions](#queries)  
4. [Summary and Key Insights](#summary)


---

### Key Questions Explored via SQL

1. What is the average CGPA by placement status?
2. How many internships do students have by placement status?
3. What is the average soft skills rating by placement status?
4. What is the placement rate by number of projects?
5. What is the average aptitude test score by placement status?
6. How many students are there by placement status, and what are their average SSC marks?
7. What is the average CGPA for each soft skills rating?
8. Which placement status group has the highest average aptitude test score?

---

### Next Steps

- Step 5: [Modeling & Prediction](./05_modeling_prediction.ipynb)

<a id="import"></a>

## **3.1 Import Libraries, Load the Dataset, and Connect to the Database**

We begin by importing the required Python libraries:

- **Pandas** for data manipulation and analysis.
- **SQLite3** for setting up and querying a local SQL database.
- **IPython SQL extension** for running SQL queries directly within the notebook.

After loading the **Student Depression** dataset into a pandas DataFrame, we connect it to a local **SQLite** database. This setup enables us to perform SQL-based exploration conveniently within the Jupyter Notebook environment.

In [1]:
# Pandas is a software library written for the Python programming language for data manipulation and analysis.
import pandas as pd

# sqlite3 is a built-in Python library for creating and interacting with SQLite databases
import sqlite3

# prettytable is a module used to format tabular data in a readable way (optional)
import prettytable

In [2]:
# Load the ipython-sql extension to run SQL queries in Jupyter
%load_ext sql

# Sets default format for prettytable display
prettytable.DEFAULT = 'DEFAULT'

In [3]:
# Create SQLite connection and cursor
con = sqlite3.connect("placement_analysis.db")
cur = con.cursor()

# Connect to the SQLite database for ipython-sql
%sql sqlite:///placement_analysis.db

In [4]:
# Load your cleaned dataset
df = pd.read_csv("placement_cleaned.csv")

# Save the DataFrame to a new table in the SQLite database
df.to_sql("PlacementAnalysis", con, if_exists='replace', index=False)

9928

In [5]:
# Drop the cleaned query table if it already exists (to avoid duplication)
%sql DROP TABLE IF EXISTS PlacementData;

 * sqlite:///placement_analysis.db
Done.


[]

In [6]:
# Create a new cleaned SQL table with all non-null rows (modify as needed)
%sql CREATE TABLE PlacementData AS SELECT * FROM PlacementAnalysis;

 * sqlite:///placement_analysis.db
Done.


[]

---

<a id="preview"></a>

## **3.2 Preview the Dataset**

To understand the structure and content of the dataset, we start by viewing the first few rows using an SQL query:

In [7]:
%%sql
SELECT * 
FROM PlacementData 
LIMIT 10;

 * sqlite:///placement_analysis.db
Done.


[(7.5, 1, 1, 1, 65, 4.4, 'No', 'No', 61, 79, 'NotPlaced'),
 (8.9, 0, 3, 2, 90, 4.0, 'Yes', 'Yes', 78, 82, 'Placed'),
 (7.3, 1, 2, 2, 82, 4.8, 'Yes', 'No', 79, 80, 'NotPlaced'),
 (7.5, 1, 1, 2, 85, 4.4, 'Yes', 'Yes', 81, 80, 'Placed'),
 (8.3, 1, 2, 2, 86, 4.5, 'Yes', 'Yes', 74, 88, 'Placed'),
 (7.0, 0, 2, 2, 71, 4.2, 'Yes', 'No', 55, 66, 'NotPlaced'),
 (7.7, 1, 1, 1, 76, 4.0, 'No', 'No', 62, 65, 'NotPlaced'),
 (7.7, 2, 1, 0, 85, 3.5, 'Yes', 'Yes', 59, 72, 'NotPlaced'),
 (6.5, 1, 1, 0, 84, 3.9, 'No', 'Yes', 75, 71, 'NotPlaced'),
 (7.8, 1, 3, 2, 79, 4.4, 'Yes', 'Yes', 85, 86, 'Placed')]

---

<a id="sql_queries"></a>

## **3.3 Run SQL Queries to Answer Key Questions**

Here we run a series of SQL queries to explore and gain insights from the dataset. Each query focuses on a specific question relevant to understanding factors related to placement status.

### 3.3.1 Average CGPA by Placement Status

**Description:**
Calculate the average CGPA for students based on their placement status (placed or not placed). This highlights how academic performance (CGPA) relates to placement outcomes.

In [8]:
%%sql
SELECT placementstatus, AVG(cgpa) AS avg_cgpa
FROM PlacementData
GROUP BY placementstatus;

 * sqlite:///placement_analysis.db
Done.


[('NotPlaced', 7.468005516290295), ('Placed', 8.01153380179307)]

### 3.3.2 Count of Internships by Placement Status

**Description:**
Count the total number of internships completed by students grouped by their placement status. This reveals whether students with more internship experience tend to be placed more often.

In [9]:
%%sql
SELECT placementstatus, SUM(internships) AS total_internships
FROM PlacementData
GROUP BY placementstatus;

 * sqlite:///placement_analysis.db
Done.


[('NotPlaced', 5231), ('Placed', 5122)]

### 3.3.3 Average Soft Skills Rating by Placement Status

**Description:**
Calculate the average soft skills rating for each placement status group. This shows if students with stronger soft skills are more likely to be placed.

In [10]:
%%sql
SELECT placementstatus, AVG(softskillsrating) AS avg_softskills
FROM PlacementData
GROUP BY placementstatus;

 * sqlite:///placement_analysis.db
Done.


[('NotPlaced', 4.174366488536459), ('Placed', 4.526338744850982)]

### 3.3.4 Placement Rate by Number of Projects

**Description:**
Determine the placement rate (percentage of students placed) grouped by the number of projects completed. This helps analyze if project work experience impacts placement outcomes.

In [11]:
%%sql
SELECT projects, placementstatus, COUNT(*) AS count
FROM PlacementData
GROUP BY projects, placementstatus
ORDER BY projects, placementstatus;

 * sqlite:///placement_analysis.db
Done.


[(0, 'NotPlaced', 12),
 (0, 'Placed', 18),
 (1, 'NotPlaced', 2947),
 (1, 'Placed', 598),
 (2, 'NotPlaced', 1750),
 (2, 'Placed', 800),
 (3, 'NotPlaced', 1092),
 (3, 'Placed', 2711)]

### 3.3.5 Average Aptitude Test Score by Placement Status

**Description:**
Calculate the average aptitude test score grouped by placement status. This helps identify whether students with higher aptitude scores are more likely to secure placements.

In [12]:
%%sql
SELECT placementstatus, AVG(aptitudetestscore) AS avg_aptitude
FROM PlacementData
GROUP BY placementstatus;

 * sqlite:///placement_analysis.db
Done.


[('NotPlaced', 75.82933976900534), ('Placed', 84.3617639932154)]

### 3.3.6 Count of Students Grouped by Placement Status and Average SSC Marks

**Description:**
Group students by placement status, count them, and calculate the average SSC marks for each group, ordering the result by the average SSC marks in descending order.

In [13]:
%%sql
SELECT placementstatus, COUNT(*) AS StudentCount, AVG(SSC_Marks) AS AverageSSCMarks
FROM PlacementData
GROUP BY placementstatus
ORDER BY AverageSSCMarks DESC;

 * sqlite:///placement_analysis.db
Done.


[('Placed', 4127, 74.85267748970196), ('NotPlaced', 5801, 64.99603516635062)]

### 3.3.7 Average CGPA for Each Soft Skills Rating

**Description:** 
Calculate the average CGPA for each soft skills rating value, and order the result by average CGPA in descending order to see which soft skills rating is associated with higher CGPA.

In [14]:
%%sql
SELECT softskillsrating, AVG(cgpa) AS AverageCGPA
FROM PlacementData
GROUP BY softskillsrating
ORDER BY AverageCGPA DESC;

 * sqlite:///placement_analysis.db
Done.


[(4.8, 8.08825910931174),
 (4.7, 7.93980463980464),
 (4.6, 7.785678704856787),
 (4.5, 7.740525114155251),
 (4.4, 7.641237113402062),
 (4.3, 7.604086845466155),
 (4.2, 7.52485549132948),
 (4.1, 7.515349544072949),
 (4.0, 7.480310880829016),
 (3.8, 7.46412037037037),
 (3.9, 7.41906779661017),
 (3.6, 7.385477178423237),
 (3.7, 7.380243161094224),
 (3.4, 7.3104),
 (3.3, 7.281927710843373),
 (3.2, 7.251020408163265),
 (3.5, 7.233689839572193),
 (3.0, 7.135),
 (3.1, 7.057142857142857)]

### 3.3.8 Subquery to Find the Placement Status with Highest Average Aptitude Score

**Description:** 
Use a subquery to calculate the average aptitude test score for each placement status, then retrieve the placement status with the highest average score.

In [15]:
%%sql
SELECT placementstatus
FROM (
    SELECT placementstatus, AVG(aptitudetestscore) AS AverageAptitudeScore
    FROM PlacementData
    GROUP BY placementstatus
) AS SubQuery
ORDER BY AverageAptitudeScore DESC
LIMIT 1;

 * sqlite:///placement_analysis.db
Done.


[('Placed',)]

In [16]:
# Commit and close the connection
con.commit()
con.close()

<a id="summary"></a>

## **3.4 Summary and Key Insights**

After running the SQL queries, we summarize the key findings and insights to guide the next phases of the project.

### **Summary**

This analysis examined key aspects of student placement outcomes using the dataset provided. We explored how various factors—such as CGPA, internships, soft skills, projects, aptitude scores, SSC marks, and soft skills ratings—relate to placement status. The queries provided quantitative insights into these relationships, helping to identify patterns that correlate with higher placement rates.

### **Key Insights**

**Average CGPA and Placement**  
Students who were placed had a higher average CGPA (**8.01**) compared to those not placed (**7.47**). This suggests that academic performance may play a role in placement outcomes.

**Internships Experience**  
Interestingly, the total number of internships was similar between placed and not placed students (~**5,200** each). This indicates that internships alone might not be a distinguishing factor.

**Soft Skills Matter**  
Placed students had a higher average soft skills rating (**4.53**) compared to not placed students (**4.17**). This highlights the potential importance of soft skills development for placement success.

**Projects and Placement Rate**  
Students with 3 projects had the highest placement rate (**2,711** placed vs. **1,092** not placed). This suggests that hands-on project work could significantly enhance placement opportunities.

**Aptitude Test Scores**  
Placed students also had higher average aptitude test scores (**84.36**) than those not placed (**75.83**). This underscores the value of strong test performance.

**SSC Marks and Placement**  
Placed students had higher average SSC marks (**74.85**) compared to not placed students (**64.99**), suggesting that consistent academic performance even at earlier stages can impact placement success.

**Soft Skills Rating and CGPA**  
There’s a consistent positive trend between higher soft skills ratings and CGPA, with ratings above **4.5** showing average CGPAs above **7.7.** This indicates that students who develop soft skills tend to maintain strong academic performance.

**Placement Status with Highest Average Aptitude Score**  
The subquery revealed that the "Placed" status had the highest average aptitude test score, reinforcing the importance of strong test performance.

### **Conclusion**

The analysis demonstrates clear trends linking placement outcomes to CGPA, soft skills, aptitude test scores, and project work. While internships alone may not significantly differentiate placement status, other factors collectively contribute to higher placement rates. These insights provide a solid foundation for developing predictive models to forecast placement outcomes based on student profiles, which will be the focus of the next step in this project.